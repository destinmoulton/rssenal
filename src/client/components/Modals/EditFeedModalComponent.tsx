import * as React from "react";

import { Button, Header, Icon, Modal } from "semantic-ui-react";
import SelectFolder from "./SelectFolder";

import { TFolderID, TFolders } from "../../types";

export interface IEditFeedModalMapDispatch {
    feedInitiateUpdate: (feedInfo: any) => void;
}

export interface IEditFeedModalMapState {
    folders: TFolders;
}

interface IEditFeedModalProps {
    isModalOpen: boolean;
    onCloseModal: () => void;
    feed: any;
}

type TAllProps = IEditFeedModalMapDispatch &
    IEditFeedModalMapState &
    IEditFeedModalProps;

interface IState {
    newFeed: any;
}

class EditFeedModalComponent extends React.Component<TAllProps, IState> {
    state: IState = {
        newFeed: { title: "" }
    };

    static getDerivedStateFromProps(props: TAllProps, state: IState) {
        if (props.feed) {
            return {
                newFeed: props.feed
            };
        }
        return null;
    }

    _handleClose = () => {
        this.setState({
            newFeed: {}
        });

        this.props.onCloseModal();
    };

    _handleSave = () => {
        const { newFeed } = this.state;

        if (newFeed.name !== "") {
            this.props.feedInitiateUpdate(newFeed);
            this._handleClose();
        }
    };

    _handleInputKeyPress = (e: any) => {
        if (e.key === "Enter") {
            this._handleSave();
        }
    };

    _handleInputOnChange = (e: React.FormEvent<HTMLInputElement>) => {
        const { newFeed } = this.state;
        newFeed.title = e.currentTarget.value;
        this.setState({
            newFeed
        });
    };

    _handleSelectFolderChange = (folderId: TFolderID) => {
        const { newFeed } = this.state;
        newFeed.folder_id = folderId;

        this.setState({
            newFeed
        });
    };

    _buildEditInput() {
        const { newFeed } = this.state;
        const { folders } = this.props;

        return (
            <div>
                <label>Feed Name</label>
                <input
                    autoFocus
                    value={newFeed.title}
                    placeholder="Feed name..."
                    onKeyPress={this._handleInputKeyPress}
                    onChange={this._handleInputOnChange}
                    className="rss-editfeed-input"
                />
                <br />
                <br />
                <label>Folder</label>
                <br />
                <SelectFolder
                    selectedValue={newFeed.folder_id}
                    onChange={this._handleSelectFolderChange}
                    folders={folders}
                />
            </div>
        );
    }

    _buildButtons() {
        return (
            <div>
                <Button
                    color="orange"
                    content="Cancel"
                    floated="left"
                    icon="cancel"
                    inverted
                    onClick={this._handleClose}
                />
                <Button
                    color="green"
                    content="Save"
                    icon="save"
                    inverted
                    onClick={this._handleSave}
                />
            </div>
        );
    }

    render() {
        const { isModalOpen } = this.props;
        const form = this._buildEditInput();

        const buttons = this._buildButtons();
        return (
            <span>
                <Modal
                    open={isModalOpen}
                    onClose={this._handleClose}
                    size="tiny"
                    className="rss-modal"
                >
                    <Header icon="pencil" content="Edit Feed" />
                    <Modal.Content>{form}</Modal.Content>
                    <Modal.Actions>{buttons}</Modal.Actions>
                </Modal>
            </span>
        );
    }
}

export default EditFeedModalComponent;
