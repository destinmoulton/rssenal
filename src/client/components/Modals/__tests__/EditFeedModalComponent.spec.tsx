import * as React from "react";

import { shallow } from "enzyme";

import EditFeedModalComponent from "../EditFeedModalComponent";

import feed from "../../../../../test/data/feed";
import folders from "../../../../../test/data/folders";

describe("<EditFeedModalComponent />", () => {
    it("renders and matches the snapshot", () => {
        const wrapper = shallow(
            <EditFeedModalComponent
                folders={folders}
                feed={feed}
                feedInitiateUpdate={jest.fn()}
                isModalOpen={true}
                onCloseModal={jest.fn()}
            />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
