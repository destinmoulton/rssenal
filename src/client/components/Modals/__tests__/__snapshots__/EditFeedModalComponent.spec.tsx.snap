// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditFeedModalComponent /> renders and matches the snapshot 1`] = `
<span>
  <Modal
    centered={true}
    className="rss-modal"
    closeOnDimmerClick={true}
    closeOnDocumentClick={false}
    dimmer={true}
    eventPool="Modal"
    onClose={[Function]}
    open={true}
    size="tiny"
  >
    <Header
      content="Edit Feed"
      icon="pencil"
    />
    <ModalContent>
      <div>
        <label>
          Feed Name
        </label>
        <input
          autoFocus={true}
          className="rss-editfeed-input"
          onChange={[Function]}
          onKeyPress={[Function]}
          placeholder="Feed name..."
          value="High Scalability"
        />
        <br />
        <br />
        <label>
          Folder
        </label>
        <br />
        <SelectFolder
          folders={
            Immutable.OrderedMap {
              "5b33c4bab2438d5708dc1967": Object {
                "__v": 0,
                "_id": "5b33c4bab2438d5708dc1967",
                "name": "Random",
                "order": 0,
              },
              "5b33c57eb2438d5708dc1968": Object {
                "__v": 0,
                "_id": "5b33c57eb2438d5708dc1968",
                "name": "Programming",
                "order": 1,
              },
            }
          }
          onChange={[Function]}
          selectedValue="5b33c57eb2438d5708dc1968"
        />
      </div>
    </ModalContent>
    <ModalActions>
      <div>
        <Button
          as="button"
          color="orange"
          content="Cancel"
          floated="left"
          icon="cancel"
          inverted={true}
          onClick={[Function]}
          role="button"
        />
        <Button
          as="button"
          color="green"
          content="Save"
          icon="save"
          inverted={true}
          onClick={[Function]}
          role="button"
        />
      </div>
    </ModalActions>
  </Modal>
</span>
`;
